ui/lib/kubernetes/addon/components/page/role/create-and-edit.js

```diff
@@ -0,0 +1,130 @@
+import { getRules } from '../../../utils/generated-role-rules';
+import { htmlSafe } from '@ember/template';
+import errorMessage from 'vault/utils/error-message';
+
+export default class CreateAndEditRolePageComponent extends Component {
+  @service router;
+  @service flashMessages;
+
+  @tracked roleRulesTemplates;
+  @tracked selectedTemplateId = '1';
+  @tracked modelValidations;
+
+  constructor() {
+    super(...arguments);
+    // first check if generatedRoleRules matches one of the templates, the user may have chosen a template and not made changes
+    // in this case we need to select the corresponding template in the dropdown
+    // if there is no match then replace the example rules with the user defined value for no template option
+    const { generatedRoleRules } = this.args.model;
+    const rulesTemplates = getRules();
+    if (generatedRoleRules) {
+      const template = rulesTemplates.findBy('rules', generatedRoleRules);
+      if (template) {
+        this.selectedTemplateId = template.id;
+      } else {
+        rulesTemplates.findBy('1').rules = generatedRoleRules;
+      }
+    }
+    this.roleRulesTemplates = rulesTemplates;
+  }
+
+  get generationPreferences() {
+    return [
+      {
+        title: 'Generate token only using existing service account',
+        description:
+          'Enter a service account that already exists in Kubernetes and Vault will dynamically generate a token.',
+        value: 'basic',
+      },
+      {
+        title: 'Generate token, service account, and role binding objects',
+        description:
+          'Enter a pre-existing role (or ClusterRole) to use. Vault will generate a token, a service account and role binding objects.',
+        value: 'expanded',
+      },
+      {
+        title: 'Generate entire Kubernetes object chain',
+        description:
+          'Vault will generate the entire chain— a role, a token, a service account, and role binding objects— based on rules you supply.',
+        value: 'full',
+      },
+    ];
+  }
+
+  get extraFields() {
+    return [
+      {
+        type: 'annotations',
+        key: 'extraAnnotations',
+        description: 'Attach arbitrary non-identifying metadata to objects.',
+      },
+      {
+        type: 'labels',
+        key: 'extraLabels',
+        description:
+          'Labels specify identifying attributes of objects that are meaningful and relevant to users.',
+      },
+    ];
+  }
+
+  get roleRulesHelpText() {
+    const message =
+      'This specifies the Role or ClusterRole rules to use when generating a role. Kubernetes documentation is';
+    const link =
+      '<a href="https://kubernetes.io/docs/reference/access-authn-authz/rbac/" target="_blank" rel="noopener noreferrer">available here</>';
+    return htmlSafe(`${message} ${link}.`);
+  }
+
+  @action
+  resetRoleRules() {
+    this.roleRulesTemplates = getRules();
+  }
+
+  @action
+  selectTemplate(event) {
```

Comment:
It may be a little cleaner to destructure `event` in the arguments so it would look like `selectTemplate({ target })` and update the parts where you use `event.target.value` accordingly.

Reply:
I like using `event` in these cases so it is clear what is being passed into the method. I could destructure `value` from `event.target` but since it's only being used once I'm not sure if it's worthwhile. What do you think?

Reply:
I agree - it's not necessary to destructure `value` for this situation :) I also saw that you'll remove this action since it's  not being used!